### ISSUE TITLE
Support configuring per-feature pkg-config names

### ISSUE THREAD

**Issue by sdroege** [gdesmott/system-deps#11]
In GStreamer the GL library was split from one big pkg-config file into separate ones per platform (EGL, X11, Wayland) plus the main one (GL).

For backwards compatibility in the crate for gstreamer-gl-x11-sys (for example) it would be needed to depend on `gstreamer-gl-1.0` for versions < 1.18 and on `gstreamer-gl-x11-1.0` for versions >= 1.18.

Similarly gstreamer-gl-x11 would have to depend on the gstreamer-gl-sys crate (with "x11" feature) for < 1.18 and on gstreamer-gl-x11-sys (without any features) for >= 1.18.

I don't think either of this can currently be expressed.

**Comment by gdesmott** [comment]
Indeed I don't think it's currently possible.

Ideally it would be nice to re-use the existing [feature versions](https://docs.rs/system-deps/1.3.2/system_deps/#feature-versions) if possible but I'm not sure how to properly express all that. :\

**Comment by sdroege** [comment]
See also https://gitlab.freedesktop.org/gstreamer/gstreamer-rs-sys/-/merge_requests/65 for the relevant MR

**Comment by sdroege** [comment]
> Ideally it would be nice to re-use the existing [feature versions](https://docs.rs/system-deps/1.3.2/system_deps/#feature-versions) if possible but I'm not sure how to properly express all that. :\

Maybe

```
feature-versions = { v1_14 = "gstreamer-gl-1.0/1.14", v1_18 = "gstreamer-gl-egl-1.0/1.18" }
```

?

**Comment by gdesmott** [comment]
A bit magic but I suppose it could work as crate names are not supposed to contain `/`.

A more explicit version way could be:
```
gst_gl = { v1_14 = { name = "gstreamer-gl-1.0", version = "1.14"} , v1_18 = { name = "gstreamer-gl-egl-1.0", version = "1.18" }}
```

**Comment by sdroege** [comment]
> A bit magic but I suppose it could work as crate names are not supposed to contain `/`.

`pkg-config` names, not crate names

**Comment by gdesmott** [comment]
Right.

Which version do you prefer? I think I'd vote for the second following the good old "explicit is better than implicit" principle.

**Comment by sdroege** [comment]
I agree

**Comment by gdesmott** [comment]
Let's go for that then.

Unfortunately I'm pretty busy these days so I'm afraid I won't have time to work on this any time soon. But if you have some time I'm happy to merge and release a patch adding this feature.

**Comment by sdroege** [comment]
Same, updating the bindings used up a lot of my spare time for this week. I'll do the hack I did for the stable version of the bindings for now then.

**Comment by gdesmott** [comment]
> A more explicit version way could be:
> 
> `gst_gl = { v1_14 = { name = "gstreamer-gl-1.0", version = "1.14"} , v1_18 = { name = "gstreamer-gl-egl-1.0", version = "1.18" }}`

This is not coherent with the existing syntax for feature version:

`gstreamer = { name = "gstreamer-1.0", version = "1.0", feature-versions = { v1_2 = "1.2", v1_4 = "1.4", v1_6 = "1.6" }}`

So we could either use the same pattern and introduce `feature-names` :  
 `gst_gl = { name = "gstreamer-gl-1.0", version = "1.14" , feature-versions = {v1_18 = "1.18" }, feature-names = { v1_18 = "gstreamer-gl-egl-1.0" }}`

That split the versioning information, so it may be best to merge them all together as suggested above:
`gst_gl = { v1_14 = { name = "gstreamer-gl-1.0", version = "1.14"} , v1_18 = { name = "gstreamer-gl-egl-1.0", version = "1.18" }}`

This syntax is more coherent and could be easily extended to override more fields later if needed.
But that would be a config syntax break, assuming we stop supporting the current syntax.

The new syntax would also be more verbose, as something currently wrote using

```toml
[package.metadata.system-deps.gstreamer_1_0]
name = "gstreamer-1.0"
version = "1.8"

[package.metadata.system-deps.gstreamer_1_0.feature-versions]
v1_10 = "1.10"
v1_12 = "1.12"
```

will have to be written using
```toml
[package.metadata.system-deps.gstreamer_1_0]
name = "gstreamer-1.0"
version = "1.8"

[package.metadata.system-deps.gstreamer_1_0.v1_10]
version = "1.10"
[package.metadata.system-deps.gstreamer_1_0.v1_12]
version = "1.12"
```

But as most of this is generated anyways maybe it's not that bad.

@sdroege : what do you think?

**Comment by sdroege** [comment]
Seems nicely clean to me.

But things like

> ```toml
> [package.metadata.system-deps.gstreamer_1_0.v1_12]
> version = "1.12"
> ```

could also be omitted in theory. We could define that a `v1_12` automatically maps to `1.12` as requirement unless otherwise specified. That way you could specify all features via an array.

**Comment by gdesmott** [comment]
Default feature -> version sounds like a good idea. But I think I'd still remove the existing `feature-versions` to not have 3 way to deal with those (the default, `version` field in a version specific dict and `feature-versions`).

The default will likely cover the existing usage so porting will be a simple matter of remove `feature-versions`.

**Comment by sdroege** [comment]
Sounds good to me, thanks :)

**Comment by gdesmott** [comment]
> I'll do the hack I did for the stable version of the bindings for now then.

@sdroege : which hack is that? I don't see any use of `gstreamer-gl-egl` in `gstreamer-rs-sys`.

**Comment by sdroege** [comment]
Yes, that's the hack. The .gir files are merged into `GstGL-1.0.gir` and only `gstreamer-gl-1.0` is used. There are WIP MRs on gstreamer-sys and gstreamer-rs that switch to the separate ones.

**Comment by gdesmott** [comment]
I just released 2.0.0 with this change.
